Sign_minion_messages support

cro opened the pull request:
### What does this PR do?

Add support for signing messages that minions `_fire_master` onto the bus.  Master will verify the signature using its copy of the minion's public key.

@cachedout , @thatch45 need your review please.

cro committed:
Sign_minion_messages support

salt-jenkins commented:
@cro, thanks for your PR! By analyzing the history of the files in this pull request, we identified @SmithSamuelM, @thatch45 and @cachedout to be potential reviewers.

cro commented:
This is still a WIP, I have some concerns about performance here.  Computing a signature twice (minion and master side) for every single event sent back from a minion could impact CPU significantly.  In addition we are naively re-reading the keys from disk every single time--some more intelligent caching could potentially mitigate some performance hit.

cachedout commented:
Go Go Jenkins!

thatch45 commented:
Reading the key off disk is the bigger hit here, the signature should not be that heavy. so yes, we should be enabling caching the minion key to make this better. Other than that I think this looks ok

cro committed:
Fat finger fix.

cro committed:
Add caching of key.

cro commented:
OK, this latest push includes some simple caching.

Note that the memoization process will leak a tiny amount of memory, but it is trivial since keys don't change that often.  On the minion side if the key changes the minion will get restarted anyway.

cro commented:
I should note that signature failures are advisory only right now.

cachedout commented:
Some interesting test failures here. I will re-run this.

cachedout commented:
Go Go Jenkins!

cro commented:
Actually I don't seem to have this quite right yet.  Something about this code is only verifying the signature on the minion_start events, not the returns.

I'm also adding the following config options and documentation for them:

```
    # Require that the minion sign messages it posts to the master on the event
    # bus
    'minion_sign_messages': bool,

    # Have master drop messages from minions for which their signatures do
    # not verify
    'drop_messages_signature_fail': bool,

    # Require that payloads from minions have a 'sig' entry
    # (in other words, require that minions have 'minion_sign_messages'
    # turned on)
    'require_minion_sign_messages': bool,
```

cro committed:
Add documentation to the example master and minion configuration files.
Move minion event signing to a saner place.
Enable dropping messages when signature does not verify or when
minion is not adding the signature to its payloads.

cro committed:
Lint

cro commented:
OK, I think that did it.  Waiting for tests now

cro commented:
Looking into that one test failure

cro committed:
Correct test--caching requires files on disk but the test just supplies
what would have been read from disk.
